pkg/controllers/cloud/networkpolicy.go

```diff
@@ -872,47 +744,235 @@ func (a *appliedToSecurityGroup) updateRules(r *NetworkPolicyReconciler) error {
 		return fmt.Errorf("unable to get networkPolicy with key %s from indexer: %w", a.id.Name, err)
 	}
 	if len(nps) == 0 {
-		if a.hasMembers {
-			r.Log.V(1).Info("AppliedToSecurityGroup clear members with no rules", "Name", a.id.Name)
-			ch := securitygroup.CloudSecurityGroup.UpdateSecurityGroupMembers(&a.id, nil, false)
-			go func() {
-				err := <-ch
-				r.cloudResponse <- &securityGroupStatus{sg: a, op: securityGroupOperationClearMembers, err: err}
-			}()
-			return nil
-		}
-		// No need to update appliedToSecurityGroup with no members.
-		r.Log.V(1).Info("AppliedToSecurityGroup clear rules", "Name", a.id.Name)
-		a.hasRules = false
+		a.clearMembers(r)
 		return nil
 	}
-	irules, erules := a.computeRules(nps)
-	if irules == nil || erules == nil {
-		return nil
+
+	for _, obj := range nps {
+		np := obj.(*networkPolicy)
+		a.updateANPRules(r, np)
+	}
+
+	return nil
+}
+
+// updateANPRules invokes cloud plug-in to update rules of appliedToSecurityGroup for a given ANP.
+func (a *appliedToSecurityGroup) updateANPRules(r *NetworkPolicyReconciler, np *networkPolicy) {
+	if !a.isReady() || a.deletePending || !np.rulesReady {
+		return
+	}
+
+	npNamespacedName := np.getNamespacedName()
+	nps, err := r.networkPolicyIndexer.ByIndex(networkPolicyIndexerByAppliedToGrp, a.id.Name)
+	if err != nil {
+		r.Log.Error(err, "get networkPolicy indexer", "sg", a.id.Name)
+		err = fmt.Errorf("internal error when updating rules for sg %s anp %s", a.id.Name, npNamespacedName)
+		r.sendRuleRealizationStatus(&np.NetworkPolicy, err)
+		a.status = err
+		_ = a.updateNPTracker(r)
+		return
+	}
+	if len(nps) == 0 {
```

Comment:
why would you have a case of 0 nps here.. put a log here (error log). it seems to me a dead code